crates/blockchain/payload.rs

```diff
@@ -145,19 +144,327 @@ fn calc_excess_blob_gas(parent_excess_blob_gas: u64, parent_blob_gas_used: u64)
+                > U256::from(MAX_BLOB_GAS_PER_BLOCK)
+        {
+            info!("No more blob gas to run blob transactions");
+            blob_txs.clear();
+        }
+        // Fetch the next transactions
+        let (head_tx, is_blob) = match (plain_txs.peek(), blob_txs.peek()) {
+            (None, None) => break,
+            (None, Some(tx)) => (tx, true),
+            (Some(tx), None) => (tx, false),
+            (Some(a), Some(b)) if compare_heads(&a, &b).is_lt() => (b, true),
+            (Some(tx), _) => (tx.clone(), false),
+        };
+
+        let txs = if is_blob {
+            &mut blob_txs
+        } else {
+            &mut plain_txs
+        };
+
+        // Check if we have enough gas to run the transaction
+        if context.remaining_gas < head_tx.tx.gas_limit() {
+            info!(
+                "Skipping transaction: {}, no gas left",
+                head_tx.tx.compute_hash()
+            );
+            // We don't have enough gas left for the transaction, so we skip all txs from this account
+            txs.pop();
+            continue;
+        }
+        // Pull transaction from the mempool
+        // TODO: maybe fetch hash too when filtering mempool so we don't have to compute it here (we can do this in the same refactor as adding timestamp)
+        let tx_hash = head_tx.tx.compute_hash();
+        mempool::remove_transaction(tx_hash, context.store())?;
+
+        // Check wether the tx is replay-protected
+        if head_tx.tx.protected() && !chain_config.is_eip155_activated(context.block_number()) {
+            // Ignore replay protected tx & all txs from the sender
+            info!("Ignoring replay-protected transaction: {}", tx_hash);
+            txs.pop();
+            continue;
+        }
+        // Execute tx
+        let receipt = match apply_transaction(&head_tx, context) {
+            Ok(receipt) => {
+                txs.shift();
+                receipt
+            }
+            // Ignore following txs from sender
+            Err(_) => {
+                info!("Failed to execute transaction: {}", tx_hash);
+                txs.pop();
+                continue;
+            }
+        };
+        // Add transaction to block
+        info!("Adding transaction: {} to payload", tx_hash);
+        context.payload.body.transactions.push(head_tx.tx);
+        // Save receipt for hash calculation
+        context.receipts.push(receipt);
+    }
+    Ok(())
+}
+
+/// Executes the transaction, updates gas-related context values & return the receipt
+// TODO: Handle blobs in blob txs
+fn apply_transaction(
+    head: &HeadTransaction,
+    context: &mut PayloadBuildContext,
+) -> Result<Receipt, ChainError> {
+    let result = execute_tx(
+        &head.tx,
+        &context.payload.header,
+        context.evm_state,
+        spec_id(context.store(), context.payload.header.timestamp)?,
+    )?;
+    context.remaining_gas -= result.gas_used();
```

Comment:
We check that the remaining gas is not lower than the transaction's gas_limit before running the transaction, so that case is not possible at the moment of applying the transaction to the payload

Reply:
Makes sense, but the check is being done in another function, right?
I'd prefer not to leave it to the chance of a refactor breaking this invariant.

Instead, can we do something like the following?
```rust
context.remaining_gas.checked_sub(result.gas_used()).ok_or(*an error here*)?;
```

Reply:
This is an internal function, it won't be used for anything other than payload building in a single `build_payload` function. Would you agree on using `saturating_sub` and adding a comment referencing the check that is done before calling the function?. I don't want to add a useless error variant to `ChainError` that we will have to map everywhere else.

Reply:
Sure, let's go with saturating_sub then 🚀